tensornetwork/contractors/custom_path_solvers/nconinterface_test.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2019 The TensorNetwork Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import pytest
+# pylint: disable=line-too-long
+from tensornetwork.contractors.custom_path_solvers.nconinterface import ncon_solver, ncon_to_adj, ord_to_ncon, ncon_cost_check
+
+
+def test_ncon_solver():
+  # test against network with known cost
+  chi = np.random.randint(2, 10)
+  u = np.random.rand(chi, chi, chi, chi)
+  w = np.random.rand(chi, chi, chi)
+  ham = np.random.rand(chi, chi, chi, chi, chi, chi)
+  tensors = [u, u, w, w, w, ham, u, u, w, w, w]
+  connects = [[1, 3, 10, 11], [4, 7, 12, 13], [8, 10, -4], [11, 12, -5],
+              [13, 14, -6], [2, 5, 6, 3, 4, 7], [1, 2, 9, 17], [5, 6, 16, 15],
+              [8, 9, -1], [17, 16, -2], [15, 14, -3]]
+  con_order, costs, is_optimal = ncon_solver(tensors, connects, max_branch=None)
+  flat_connects = np.concatenate(connects)
+  inds = np.sort(np.unique(flat_connects[flat_connects > 0]))
+  ex_cost = np.log10(2 * chi**9 + 4 * chi**8 + 2 * chi**6 + 2 * chi**5)
+  assert np.allclose(costs, ex_cost)
+  assert is_optimal
+  assert np.array_equal(inds, np.sort(con_order))
+
+
+def test_ncon_solver2():
+  # test with random network
+  chi = np.random.randint(2, 5)
+  A = np.random.rand(chi, chi, chi, chi, chi, chi)
+  N = np.random.randint(2, 9)
+  tensors = [A] * N
+  num_closed = np.random.randint(0, 2 * N) + 1
+  num_open = 4 * N - 2 * num_closed
+  cl_inds = 1 + np.arange(num_closed)
+  op_inds = -1 - np.arange(num_open)
+  connects = [0] * N
+  rand_perm = np.argsort(np.random.rand(4 * N))
+  comb_inds = np.concatenate((op_inds, cl_inds, cl_inds))[rand_perm]
+  for k in range(N):
+    if k < (N - 1):
+      connect_temp = np.concatenate((comb_inds[4 * k:4 * (k + 1)],
+                                     [num_closed + k + 1, num_closed + k + 2]))
+    else:
+      connect_temp = np.concatenate(
+          (comb_inds[4 * k:4 * (k + 1)], [num_closed + k + 1, num_closed + 1]))
+    connects[k] = list(connect_temp[np.argsort(np.random.rand(6))])
+  max_branch = np.random.randint(1, 1000)
```

Comment:
Try to avoid using randomized values in unit tests, especially ones that change the code path taken like this one would do.

You can either just hard code it or add a parametrize variable to this test. 
https://docs.pytest.org/en/2.8.7/parametrize.html

Reply:
This happens several times.

Reply:
Sure, I can replace the randomized networks with some instances of hard-coded networks running over some parameters. 

I was using the randomized networks for my own purposes to test that the algorithms themselves are functioning correctly (running thousands of times) to check for bugs which may only show up in rare instances (and would not be found in hard-coded tests). Given that I may want to modify the algorithms in the future, I would find it useful to keep these randomized tests for my own use. Is there a proper place for this type of test, or should I just keep this in my private repo?

Reply:
Generally you want to avoid randomized testing like that because if it DOES find an error, it can be difficult and frusterating to replicate. If you want through testing like that, just parameterize the seed of the RNG, that was it can be reproduced if an error pops up.